graftm/taxonomy.py

```diff
@@ -0,0 +1,176 @@
+    pass
+
+#-------------------------------------------------------------------------------
+
+class Taxonomy:
+
+    TAB='\t'
+    COM=','
+    GG_SEP_0='; '
+    GG_SEP_1=';'
+    
+    def __init__(self):
+        self.taxonomy = {}
+        
+    def _ammend_taxonomy_gaps(self):
+        '''
+        Checks that each value in the self.taxonomy dict has consistent
+        length. 
+        '''
+        try:
+            self.max = max([len(x) for x in self.taxonomy.values()])
+            for taxa_id, taxa_taxonomy_list in self.taxonomy.iteritems():
+                
+                if len(taxa_taxonomy_list) == self.max:
+                    continue
+                else:
+                    ammended_split_taxonomy = \
+                                taxa_taxonomy_list + \
+                                [''] * (self.max - len(taxa_taxonomy_list))
+                self.taxonomy[taxa_id] = ammended_split_taxonomy
+        except:
+            logging.error("Controlled exit from taxonomy.Taxonomy \
+ _ammend_taxonomy_gaps")    
+            ProgrammingError("Error when trying to interpret taxonomy input")
+    
+    def check(self, taxa_id):
+        '''
+        Check whether an id is within the taxonomy description.
+        
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        '''
+        if taxa_id in self.id_set:
+            return True
+        else:
+            return False
+        
+    def entry_ids(self):
+        '''
+        Return a list of all taxonomy entries within the self.taxnomy dict
+        '''
+        return self.taxonomy.keys()
+    
+    def rank(self, taxa_id, rank):
+        '''
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        rank : int
+            rank to return
+        Returns
+        -------
+        rank as indicated by the index passed to 'rank'.
+        '''
+        if rank < self.max:
+            return self.taxonomy[taxa_id][rank]
+        else:
+            raise ProgrammingError("The rank specified was larger than the \
+maximum possible number of ranks within the taxonomy description.")
+        
+        
+    
+    def taxonomy_string(self, taxa_id):
+        '''
+        Return the tax string of the id provided
+        
+        Parameters
+        ----------
+        taxa_id : array
+            taxa id to return the taxonomy for
+        
+        Returns
+        -------
+        The taxonomy string joined with the standard self.GG_SEP_0 separator.
+        '''
+        
+        if taxa_id in self.taxonomy:
+            taxonomy_string = GG_SEP_0.join(self.taxonomy[taxa_id])
+        else:
+            raise MissingTaxonomy("Taxa id provided did not have any \
+any corresponding entry in the taxonomy description.")
+            
+        return taxonomy_string
+            
+    def read_format_greengenes(self, taxonomy_io):
```

Comment:
I think it might be better to make these methods `@staticmethod`s which return a new Taxonomy instance so you can `taxonomy = Taxonomy.read_format_greengenes(io)` yeh?